File: longestCommonSubsequence.js
// Time complexity: O(2^n)
// Space complexity: O(n) due to the call stack
const LCSNaive = (text1, text2) => {
  const m = text1.length;
  const n = text2.length;
  if (!m || !n) return 0;
  if (text1[m - 1] === text2[n - 1])
    return 1 + LCSNaive(text1.slice(0, m - 1), text2.slice(0, n - 1));
  else
    return Math.max(
      LCSNaive(text1, text2.slice(0, n - 1)),
      LCSNaive(text1.slice(0, m - 1), text2)
    );
};

// Time complexity: O(mn)
// Space complexity: O(mn)
const LCSDP = (text1, text2) => {
  const m = text1.length;
  const n = text2.length;
  let ans = 0;
  const dp = [...Array(m + 1)].map(() => Array(n + 1).fill(0));
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      dp[i][j] =
        text1[i - 1] === text2[j - 1]
          ? 1 + dp[i - 1][j - 1]
          : Math.max(dp[i][j - 1], dp[i - 1][j]);
      ans = Math.max(ans, dp[i][j]);
    }
  }
  return ans;
};

File: longestCommonSubsequence.py
class Solution1:
    def LCS(self, text1: str, text2: str) -> int:
        m = len(text1)
        n = len(text2)
        if m == 0 or n == 0:
            return 0
        if text1[m - 1] == text2[n - 1]:
            return 1 + self.LCS(text1[:m - 1], text2[:n - 1])
        else:
            return max(
                self.LCS(text1[:m - 1], text2),
                self.LCS(text1, text2[:n - 1])
            )


class Solution2:
    def LCS(self, text1: str, text2: str) -> int:
        m = len(text1)
        n = len(text2)
        dp = [[None for _ in range(n + 1)] for _ in range(m + 1)]
        for i in range(m + 1):
            for j in range(n + 1):
                if i == 0 or j == 0:
                    dp[i][j] = 0
                elif text1[i - 1] == text2[j - 1]:
                    dp[i][j] = 1 + dp[i - 1][j - 1]
                else:
                    dp[i][j] = max(
                        dp[i - 1][j],
                        dp[i][j - 1]
                    )
        return dp[m][n]
